src/contracts/libraries/GPv2Signing.sol

```diff
@@ -167,6 +169,99 @@ library GPv2Signing {
+    ///     bytes signature;
+    /// }
+    /// ```
+    ///
+    /// All entries are tightly packed together in this order in the encoded
+    /// calldata. `signatureLength` encodes the length of the signature bytes
+    /// in bytes. Example:
+    ///
+    /// input:    0x73c14081446bd1e4eb165250e826e80c5a523783000a00010203040506070809
+    /// decoding:   [..............verifier................][sL][.......data.......]
+    /// stride:                                          20   2         10 (0x000a)
+    ///
+    /// This function enforces that the encoded data stores enough bytes to
+    /// cover the full length of the decoded interaction.
+    ///
+    /// The size of `dataLength` limits the maximum calldata that can be used in
+    /// a signature to 2**16 ≈ 65 kB.
+    function recoverEip1271Signer(
+        bytes calldata encodedSignature,
+        bytes32 domainSeparator,
+        bytes32 orderDigest
+    ) internal view returns (address owner, bytes calldata remainingCalldata) {
+        uint256 signatureLength;
+        bytes calldata signature;
+
+        // NOTE: Use assembly to efficiently decode signature data.
+        // If reading calldata out of bound, the extra bytes are set to zero.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // owner = address(encodedSignature[0:20])
+            owner := shr(96, calldataload(encodedSignature.offset))
+            // signatureLength = uint256(encodedSignature[20:22])
+            signatureLength := shr(
+                240,
+                calldataload(add(encodedSignature.offset, 20))
+            )
+        }
+
+        // Safety: dataLength fits a uint16 by construction, no overflow is
+        // possible.
+        uint256 usedCalldataStride = 20 + 2 + signatureLength;
+        require(
+            encodedSignature.length >= usedCalldataStride,
+            "GPv2: invalid signature encoding"
+        );
+
+        // NOTE: Use assembly to efficiently decode signature data and assign
+        // calldata skipping bounds checks.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // signature = bytes(encodedTrade[22:22+signatureLength])
+            signature.offset := add(encodedSignature.offset, 22)
+            signature.length := signatureLength
+            // remainingCalldata = bytes(encodedTrade[22+signatureLength:])
+            remainingCalldata.offset := add(
+                encodedSignature.offset,
+                usedCalldataStride
+            )
+            remainingCalldata.length := sub(
+                encodedSignature.length,
+                usedCalldataStride
+            )
+        }
+
+        uint256 freeMemoryPointer = getFreeMemoryPointer();
+
+        // The digest is chosen to be consistent with EIP-191. Its format is:
+        // 0x19 <1 byte version> <version specific data> <data to sign>.
+        // Version number 0x2a is chosen arbitrarily so that it does not
```

Comment:
Your chance to make it a standard (if other teams haven't already)!

Reply:
Maybe we should implement @rmeissner 's suggestion #155 so that we compute `keccak256(abi.encodePacked("\x19\x01", domainSeparator, orderDigest))` as a signing digest for all signature types. The downside is that `recoverEthSign` path does a double `keccak256` but I don't expect it to be used too much. My worry is the non-standard nature of `"\x19\x2a"`.

Otherwise, I love the choice of version number :smile:

Reply:
We can do that easily, it's enough to change the version bit and then we'd be using EIP712 message encoding for EIP1271 signatures, no extra cost. I avoided that because the ERC712 technically also specifies the signing scheme to be used, so this would be in principle an abuse of standard and might be confusing. On the other hand, I'd have liked for EIP712 to just specify the encoding of the message to be signed, so that different schemes could be used (and shown on an interface).
No strong opinion here, I'll change the byte if there is no further discussion.

Reply:
>  avoided that because the ERC712 technically also specifies the signing scheme to be used,

This is a good point. Lets leave it as is, and maybe open an issue so we don't forget to think about it again. It might be worth asking the Safe team what they think about it.